Implement optimized support for vector I/O in Subfiling VFD

jhendersonHDF opened the pull request:
Vector I/O requests are now processed within a single set of I/O call batches, rather than each I/O vector entry (tuple constructed from the types, addrs, sizes and bufs arrays) being processed individually. This allows I/O to be more efficiently parallelized among the I/O concentrator processes during large I/O requests.

jhendersonHDF committed:
Implement optimized support for vector I/O in Subfiling VFD

Vector I/O requests are now processed within a single
set of I/O call batches, rather than each I/O vector
entry (tuple constructed from the types, addrs, sizes
and bufs arrays) being processed individually. This allows I/O to be
more efficiently parallelized among the I/O concentrator processes
during large I/O requests.

jhendersonHDF commented on src/H5FDsubfiling/H5FDioc.c:
The changes in this file mostly just add the size-extending feature for vector I/O to the IOC VFD. The `types` parameter is not used here, so that extending feature wasn't added.

jhendersonHDF commented on src/H5FDsubfiling/H5FDioc_int.c:
Now that vector I/O is supported, vector I/O requests could get passed directly to the IOC VFD if there is only 1 subfile since it doesn't make since to split up the I/O requests across subfiles in that case. The change in this file addresses a new problem that came up with the new support for vector I/O.

jhendersonHDF commented on src/H5FDsubfiling/H5FDioc_threads.c:
Same comment here about addressing vector I/O requests getting passed directly down to the IOC VFD

jhendersonHDF commented on testpar/t_subfiling_vfd.c:
```diff
@@ -40,7 +40,7 @@
 #endif
 
-#define DEFAULT_DEFLATE_LEVEL 9
+#define DEFAULT_DEFLATE_LEVEL 4
```
Change the compression level from 9 to something a bit less aggressive since the tests were spending ~60% of the time just doing compression and that isn't the main focus for the tests.

jhendersonHDF commented on testpar/t_subfiling_vfd.c:
```diff
@@ -2360,11 +2360,33 @@ main(int argc, char **argv)
     if (MAINPROCESS)
         puts("");
 
+    if (MAINPROCESS)
```
Move the block that sets compression up so that the tests run once with compression before setting environment variables and once after setting them. Previously, the compression tests were running twice only after the environment variables had been set.

jhendersonHDF commented on src/H5FDsubfiling/H5FDsubfiling.c:
```diff
@@ -1535,199 +1565,16 @@ static herr_t
     assert(buf);
 
-    /* Check for overflow conditions */
-    if (!H5_addr_defined(addr))
```
All the logic between read and write was mostly duplicated, so I moved it into `H5FD__subfiling_io_helper`

jhendersonHDF commented on src/H5FDsubfiling/H5FDsubfiling.c:
```diff
@@ -2086,98 +1709,27 @@ H5FD__subfiling_read_vector(H5FD_t *_file, hid_t dxpl_id, uint32_t count, H5FD_m
-    /* TODO: setup real support for vector I/O */
     if (file_ptr->fa.require_ioc) {
-
-        bool       extend_sizes = false;
```
All this logic is now moved into the part of `H5FD__subfiling_io_helper` that handles the vector I/O requests.

jhendersonHDF commented on src/H5FDsubfiling/H5FDsubfiling.c:
```diff
@@ -2581,138 +2043,913 @@ H5FD__subfiling_ctl(H5FD_t *_file, uint64_t op_code, uint64_t flags, const void
+         * Generate the types, addrs, sizes and bufs I/O vectors for
+         * this I/O request.
+         */
+        status = generate_io_vectors(
```
The old `init_indep_io` function has been replaced by `generate_io_vectors`, which can now handle translating more than one (offset, I/O size, type, buffer) tuple into a set of I/O vectors that spans the subfiles. The new function is also a bit more efficient in that it directly generates the I/O vectors rather than generating lists of offset, size, types, buffer arrays that are used to populate I/O vectors later like before.

jhendersonHDF commented on src/H5FDsubfiling/H5FDsubfiling.c:
```diff
@@ -2581,138 +2043,913 @@ H5FD__subfiling_ctl(H5FD_t *_file, uint64_t op_code, uint64_t flags, const void
+ *-------------------------------------------------------------------------
+ */
+static herr_t
+translate_io_req_to_iovec(subfiling_context_t *sf_context, size_t iovec_idx, size_t iovec_len,
```
Much of the logic in this function is unchanged from the previous function it was in (`init_indep_io`). The main difference is that the buffer indexing had to change since the I/O vectors used for the final I/O requests are populated directly here rather than generating arrays that the vectors would have later been populated with.

jhendersonHDF commented on src/H5FDsubfiling/H5FDsubfiling.c:
```diff
@@ -3020,20 +3311,21 @@ init_indep_io(subfiling_context_t *sf_context, int64_t file_offset, size_t io_ne
-iovec_fill_first(subfiling_context_t *sf_context, int64_t iovec_depth, int64_t target_datasize,
-                 int64_t start_mem_offset, int64_t start_file_offset, int64_t first_io_len,
-                 int64_t *mem_offset_out, int64_t *target_file_offset_out, int64_t *io_block_len_out)
+iovec_fill_first(subfiling_context_t *sf_context, size_t iovec_len, int64_t cur_iovec_depth,
```
The changes in the `iovec_fill_` functions below are just changes to the buffer indexing for generating I/O vectors directly.

fortnern commented on src/H5FDsubfiling/H5FDsubfiling.c:
```diff
@@ -2581,138 +2043,913 @@ H5FD__subfiling_ctl(H5FD_t *_file, uint64_t op_code, uint64_t flags, const void
+                }
+
+                H5_CHECK_OVERFLOW(size, size_t, int);
+                if (MPI_SUCCESS != MPI_Bcast(bufs[i].vp, (int)size, MPI_BYTE, 0, file_ptr->comm))
```
Maybe not the focus of this PR but how hard would it be to accumulate these into a single bcast?

jhendersonHDF replied:
Indeed, this was mostly just copied over from the previous code, but it could probably be handled better with a derived type instead of multiple bcasts.

fortnern commented on src/H5FDsubfiling/H5FDsubfiling.c:
```diff
@@ -2581,138 +2043,913 @@ H5FD__subfiling_ctl(H5FD_t *_file, uint64_t op_code, uint64_t flags, const void
+            if (!extend_types) {
+                if ((i > 0) && (types[i] == H5FD_MEM_NOLIST)) {
+                    extend_types = true;
+                    type         = types[i - 1];
```
Seems silly to use the type variable for type here when it can only be one thing, though I guess it makes the code more similar to other places. Not that it makes a huge difference

jhendersonHDF replied:
For now, I've just left this the same since it matches other places and was easier for me to reason about.

fortnern commented on src/H5FDsubfiling/H5FDsubfiling.c:
```diff
@@ -2581,138 +2043,913 @@ H5FD__subfiling_ctl(H5FD_t *_file, uint64_t op_code, uint64_t flags, const void
+
+    for (size_t i = 0; i < count; i++) {
+        if (!extend_types) {
+            if ((i > 0) && (types[i] == H5FD_MEM_NOLIST)) {
```
I don't think we need to keep scanning the list after we find H5FD_MEM_NOLIST

fortnern commented on src/H5FDsubfiling/H5FDsubfiling.c:
```diff
@@ -2581,138 +2043,913 @@ H5FD__subfiling_ctl(H5FD_t *_file, uint64_t op_code, uint64_t flags, const void
+            cur_max_num_subfiles = 2;
+        }
+        else {
+            cur_max_num_subfiles = data_size / stripe_size;
```
I'm not sure this calculation is correct. For example if the block starts at a strip boundary, the stripe size is 1000, and the data size is 2001, this will return 2 subfiles, when it really touches 3. And I think it needs to take into account the actual starting position, not just whether it's on a boundary, to avoid wasting space.

I think something like `(((cur_file_offset % stripe_size) + data_size - 1) / stripe_size) + 1` could work

jhendersonHDF replied:
The below `if (cur_file_offset % stripe_size)` is responsible for adding 1 to the value when the offset is aligned to a stripe boundary. I was less concerned about wasting space since an extra entry in the arrays would make very little difference.

fortnern replied:
In the case above this code results in an array that's too small

jhendersonHDF replied:
I see that now. I couldn't immediately find a case where the above suggestion is problematic, so I'll switch to that instead.

fortnern commented on src/H5FDsubfiling/H5FDsubfiling.c:
```diff
@@ -2792,212 +3027,268 @@ init_indep_io(subfiling_context_t *sf_context, int64_t file_offset, size_t io_ne
-            /* Account for subfiles with uniform segments */
-            if (!is_first && !is_last) {
-                bool thin_uniform_section = false;
+            if (last_subfile_idx < first_subfile_idx) {
```
Should this be else if? I don't think both could be true.

jhendersonHDF replied:
Yes (or just an `else`), it was just easier for me to read it this way at the time of writing.

fortnern commented on src/H5FDsubfiling/H5FDsubfiling.c:
```diff
@@ -2792,212 +3027,268 @@ init_indep_io(subfiling_context_t *sf_context, int64_t file_offset, size_t io_ne
-                    if (num_full_stripes)
-                        num_full_stripes--;
-                }
+                thin_uniform_section = (subfile_idx > first_subfile_idx) && (subfile_idx > last_subfile_idx);
```
Wouldn't this also be the case if it was below both? Also don't you only need to check if it's greater than last_subfile_idx? (And if it's less than first_subfile_idx in the case I mentioned)

jhendersonHDF replied:
> Wouldn't this also be the case if it was below both?

After thinking about this for a while and drawing some diagrams, I believe this is the case, yes, though it should just populate the I/O vectors with data that isn't used. Let me write a test for this case and verify, then I'll fix this.

> Also don't you only need to check if it's greater than last_subfile_idx?

Ah, yes that follows from the logic of `if (last_subfile_idx >= first_subfile_idx)`. Good catch.

fortnern commented on src/H5FDsubfiling/H5FDsubfiling.c:
```diff
@@ -2792,212 +3027,268 @@ init_indep_io(subfiling_context_t *sf_context, int64_t file_offset, size_t io_ne
+                 * subfile with the first data segment and the current
+                 * subfile index falls between the two.
+                 */
+                thin_uniform_section = thin_uniform_section || ((last_subfile_idx < subfile_idx) &&
```
Related to the above comment - I don't think you need the || here

jhendersonHDF replied:
Agreed. This was likely a leftover from fixing a bug a long time ago.

fortnern commented:
Done reviewing, looks good aside from my comments. We probably don't need to implement the more efficient rank 0 bcast right now but we should make a note of it.

jhendersonHDF committed:
Fix some calculations and add test cases for issues spotted from review

jhendersonHDF commented:
@fortnern I believe I've addressed the most important comments from your review. I've added test cases that test for all the cases discussed in the comments; the newly-added max number of subfiles calculation was indeed problematic, but the test passes after your suggested change. The "thin uniform section" calculation wasn't exactly problematic, but I added a test for it anyway just to be sure.

I haven't implemented the performance improvement for the rank0 bcast strategy yet, but I can add a comment to the code about that or document it somewhere else if you think that's better.

jhendersonHDF committed:
sp.

jhendersonHDF committed:
Remove a variable that was compensating for previous miscalculations

fortnern commented:
I think we can just create an issue for the rank 0 bcast improvement for now
